### DISCUSSION TITLE
Functional TypeScript #1: Algebraic Data Types

### DISCUSSION THREAD

**Discussion by giscus** [srijan-paul/srijan-paul.github.io#9] [Announcements]
# Functional TypeScript #1: Algebraic Data Types

Introduction to typeclasses, ADTs, and HKTs TypeScript

https://injuly.in/blog/ts-adt/

**Comment by citizen428** [comment]
I'd be very curious about a post showing the Haskell-inspired pattern you use in Go.

**Reply by srijan-paul** [reply]
I'll leave an example in this reply.

It's much more awkward in Go. I almost always use interfaces.
There are few exceptions where it's worth it – AST nodes is one.
In fact, I think the go compiler itself uses something like this.

Here is the code from this post ported to Golang (note the dispatch on `.(type)!):

```go
type UiEvent interface { isEvent() }

type Click struct { x float64; y float64 }
type Keypress struct{ code int }
type Scroll struct{}

// "Implement" the interface. Ensures no other type matches `UiEvent`
func (_ Click) isEvent()    {}
func (_ Keypress) isEvent() {}
func (_ Scroll) isEvent()   {}

// "pattern match" on UiEvent
func stringifyEvent(ev UiEvent) string {
	switch e := ev.(type) {
	case Click:
		return fmt.Sprintf("clicked %f %f", e.x, e.y)
	case Keypress:
		return fmt.Sprintf("pressed %d", e.code)
	case Scroll:
		return "Scrolled"
	}

	panic("...yeah. No pattern exhaustiveness checking")
}

func main() {
	evClick := Click{100, 30}
	evKeyboard := Keypress{10}

	fmt.Println(stringifyEvent(evClick))
	fmt.Println(stringifyEvent(evKeyboard))
}

```

Of course, the idiomatic Go solution for a simple problem like this would be:

```go
type UiEvent interface {
  stringify();
}
```
